masonry/src/widget/window_handle.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright 2024 the Xilem Authors and the Druid Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! A titlebar area widget.
+
+use accesskit::{NodeBuilder, Role};
+use dpi::LogicalPosition;
+use smallvec::{smallvec, SmallVec};
+use tracing::{trace_span, Span};
+use vello::kurbo::Point;
+use vello::Scene;
+
+use crate::event::PointerButton;
+use crate::widget::{WidgetMut, WidgetPod};
+use crate::{
+    AccessCtx, AccessEvent, BoxConstraints, EventCtx, LayoutCtx, LifeCycleCtx, PaintCtx,
+    PointerEvent, Size, StatusChange, TextEvent, Widget, WidgetId,
+};
+
+// TODO: Maybe make this configurable somehow
+const DRAG_THRESHOLD: f64 = 8.0;
+
+/// A titlebar area widget.
+///
+/// An area that can be dragged to move the window.
+pub struct WindowHandle<W: Widget> {
+    child: Option<WidgetPod<W>>,
+    last_pos: Option<LogicalPosition<f64>>,
+}
+
+// --- MARK: BUILDERS ---
+impl<W: Widget> WindowHandle<W> {
+    pub fn new(child: W) -> Self {
+        Self {
+            child: Some(WidgetPod::new(child)),
+            last_pos: None,
+        }
+    }
+
+    #[doc(alias = "null")]
+    pub fn empty() -> Self {
+        Self {
+            child: None,
+            last_pos: None,
+        }
+    }
+}
+
+// --- MARK: WIDGETMUT ---
+impl<W: Widget> WidgetMut<'_, WindowHandle<W>> {
+    pub fn set_child(&mut self, child: W) {
+        if let Some(child) = self.widget.child.take() {
+            self.ctx.remove_child(child);
+        }
+        self.widget.child = Some(WidgetPod::new(child));
+        self.ctx.children_changed();
+        self.ctx.request_layout();
+    }
+
+    pub fn remove_child(&mut self) {
+        if let Some(child) = self.widget.child.take() {
+            self.ctx.remove_child(child);
+        }
+    }
+}
+
+// --- MARK: IMPL WIDGET ---
+impl<W: Widget> Widget for WindowHandle<W> {
+    fn on_pointer_event(&mut self, ctx: &mut EventCtx, event: &PointerEvent) {
+        match event {
+            PointerEvent::PointerDown(button, state) => {
+                if !ctx.is_disabled() {
+                    // TODO: Add double click to maximize
+                    match button {
+                        PointerButton::Primary => self.last_pos = Some(state.position),
+                        PointerButton::Secondary => ctx.show_window_menu(state.position),
+                        _ => (),
+                    }
+                }
+            }
+            PointerEvent::PointerMove(state) => {
+                if let Some(last_pos) = self.last_pos {
+                    let distance = ((state.position.x - last_pos.x).powi(2)
+                        + (state.position.y - last_pos.y).powi(2))
+                    .sqrt();
```

Comment:
No need to sqrt here, since we're just doing a comparison - we can have `DRAG_THRESHOLD_SQUARED`